### PR TITLE
Issue 18 - Use a behavior to enable VLibras News API

### LINKED ISSUE #18: Create a behavior to mark that video uses VLibras
This change is to avoid GET requisitions when open old already published content.

### PR DESCRIPTION
**Pull request by rodfersou**
closes #18

### CODE REVIEW

File: src/brasil/gov/vlibrasnews/behaviors/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@provider(IFormFieldProvider)
+class IVLibrasNews(model.Schema):
+
+    """Enable VLibras News API behavior."""
+
+    model.fieldset('settings', fields=['video_url'])
+
+    video_url = schema.ASCIILine(
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/behaviors/interfaces.py]: @rodfersou use `schema.URI()` instead.

Comment (rodfersou) [reply src/brasil/gov/vlibrasnews/behaviors/interfaces.py]:
```
Module: brasil.gov.vlibrasnews.tests.test_subscribers

Traceback (most recent call last):
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/tests/test_subscribers.py", line 2, in <module>
    from brasil.gov.vlibrasnews import subscribers
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/subscribers.py", line 2, in <module>
    from brasil.gov.vlibrasnews.behaviors.interfaces import IVLibrasNews
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/behaviors/interfaces.py", line 10, in <module>
    class IVLibrasNews(model.Schema):
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/behaviors/interfaces.py", line 21, in IVLibrasNews
    required=False
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 306, in __init__
    super(MinMaxLen, self).__init__(**kw)
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 157, in __init__
    self.default = default
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 53, in __set__
    inst.validate(value)
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 182, in validate
    self._validate(value)
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_field.py", line 676, in _validate
    raise InvalidURI(value)
InvalidURI


Module: brasil.gov.vlibrasnews.tests.test_viewlets

Traceback (most recent call last):
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/tests/test_viewlets.py", line 2, in <module>
    from brasil.gov.vlibrasnews.browser.vlibrasnews import VLibrasNewsViewlet
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/browser/vlibrasnews.py", line 3, in <module>
    from brasil.gov.vlibrasnews.behaviors.interfaces import IVLibrasNews
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/behaviors/interfaces.py", line 10, in <module>
    class IVLibrasNews(model.Schema):
  File "/srv/brasil.gov.vlibrasnews/src/brasil/gov/vlibrasnews/behaviors/interfaces.py", line 21, in IVLibrasNews
    required=False
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 306, in __init__
    super(MinMaxLen, self).__init__(**kw)
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 157, in __init__
    self.default = default
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 53, in __set__
    inst.validate(value)
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_bootstrapfields.py", line 182, in validate
    self._validate(value)
  File "/srv/cache/eggs/zope.schema-4.2.2-py2.7.egg/zope/schema/_field.py", line 676, in _validate
    raise InvalidURI(value)
InvalidURI
```

Comment (rodfersou) [reply src/brasil/gov/vlibrasnews/behaviors/interfaces.py]: let's keep `schema.ASCIILine` please, the validation don't accept empty string or `None`

File: src/brasil/gov/vlibrasnews/profiles/testfixture/types.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0"?>
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/profiles/testfixture/types.xml]: @rodfersou there is no need for this fake content type; use the ones on plone.app.contenttypes

File: src/brasil/gov/vlibrasnews/testing.py [hunk 3]
```diff
@@ -31,9 +31,11 @@ class Fixture(PloneSandboxLayer):
     def setUpZope(self, app, configurationContext):
         import brasil.gov.vlibrasnews
         self.loadZCML(package=brasil.gov.vlibrasnews)
+        self.loadZCML(package=brasil.gov.vlibrasnews, name='testing.zcml')
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/testing.py]: @rodfersou always do:

```
from plone.app.contenttypes.testing import PLONE_APP_CONTENTTYPES_FIXTURE as PLONE_FIXTURE
```

File: src/brasil/gov/vlibrasnews/tests/test_controlpanel.py [hunk 4]
```diff
@@ -63,19 +62,14 @@ def test_access_token_record_in_registry(self):
...
     def test_records_removed_on_uninstall(self):
         qi = self.portal['portal_quickinstaller']
 
         with api.env.adopt_roles(['Manager']):
             qi.uninstallProducts(products=[PROJECTNAME])
 
         records = [
-            IVLibrasNewsSettings.__identifier__ + '.access_token',
-            IVLibrasNewsSettings.__identifier__ + '.enabled_content_types'
+            IVLibrasNewsSettings.__identifier__ + '.access_token'
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/tests/test_controlpanel.py]: @rodfersou don't remove the `,`

File: src/brasil/gov/vlibrasnews/tests/test_subscribers.py [hunk 5]
```diff
@@ -26,7 +24,7 @@ def setUp(self):
         with HTTMock(vlibras_ok):
             with api.env.adopt_roles(['Manager']):
                 self.document = api.content.create(
-                    type='Document',
+                    type='Dexterity Item',
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/tests/test_subscribers.py]: @rodfersou use the standard News Item for this

File: src/brasil/gov/vlibrasnews/tests/test_viewlets.py [hunk 6]
```diff
@@ -27,7 +27,7 @@ def setUp(self):
         with HTTMock(vlibras_ok):
             with api.env.adopt_roles(['Manager']):
                 self.document = api.content.create(
-                    type='Document',
+                    type='Dexterity Item',
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/tests/test_viewlets.py]: @rodfersou use the standard News Item

File: buildout.cfg [hunk 7]
```diff
@@ -15,7 +15,7 @@ parts +=
     rebuild_i18n-sh
 
 [code-analysis]
-recipe = plone.recipe.codeanalysis[recommended]
```
Comment (hvelarde) [inline buildout.cfg]: @rodfersou don't do this, pin the plugin that is failing to the version that was working before.

File: src/brasil/gov/vlibrasnews/subscribers.py [hunk 8]
```diff
@@ -162,15 +168,18 @@ def get_video_url(context):
...
-    if response.status_code != 200:
-        logger.error(u'GET - {0}: {1} - {2}'.format(
-            context.absolute_url(), response.status_code, response.reason))
-        if response.status_code != 102:
-            raise NotProcessingError('Video not ready and not being processed.')
+    if response.status_code == 200:
+        data = response.json()
+        context.video_url = data['url_youtube']
+        return
+    logger.error(u'GET - {0}: {1} - {2}'.format(
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/subscribers.py]: @rodfersou `102` status code is not an error

**Comment by hvelarde** [comment]
@rodfersou add a couple of paragraphs in the README file explaining how the behavior works.

**Comment by rodfersou** [comment]
@hvelarde done

**Comment by hvelarde** [comment]
@rodfersou thanks, but what I really want is a paragraph how the behavior works, not how to enable it.

**Comment by rodfersou** [comment]
@hvelarde I'm sorry, I don't know what more to write about the behavior.. I'm already saying that it adds a field where the video url is saved.. what else could I write about it?

**Comment by hvelarde** [comment]
never mind, I'll do it.

**Comment by rodfersou** [comment]
thank you for your wise advise...

**Comment by hvelarde** [comment]
BTW, take a look at this package: https://github.com/tallus/python-civicrm/blob/master/pythoncivicrm/pythoncivicrm.py

it shows how to write a module to access an API in a sane way; we need to do things like this, next time.

File: src/brasil/gov/vlibrasnews/subscribers.py [hunk 9]
```diff
@@ -126,6 +129,10 @@ def create_translation(context, event):
     :rtype: bool
     :raises: VLibrasNewsError
     """
+    # Unpublish item
+    if event.action != 'publish':
+        return delete_translation(context, event)
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/subscribers.py]: IMO, calling a subscriber inside another subscriber is not a good implementation; just another reason to move all the API logic outside.

this feature must be completely rewritten.

File: src/brasil/gov/vlibrasnews/tests/test_subscribers.py [hunk 10]
```diff
@@ -40,23 +41,23 @@ def test_validate(self):
 
     def test_create_translation_ok(self):
         with HTTMock(vlibras_ok):
-            self.assertTrue(subscribers.create_translation(self.document, None))
+            self.assertTrue(subscribers.create_translation(self.document, Mock(action='publish')))
```
Comment (hvelarde) [inline src/brasil/gov/vlibrasnews/tests/test_subscribers.py]: this is not how a subscriber must be tested; you should fire the event instead.

for some code examples see: https://github.com/collective/collective.fingerpointing/blob/master/src/collective/fingerpointing/tests/test_pas_subscriber.py

Comment (rodfersou) [reply src/brasil/gov/vlibrasnews/tests/test_subscribers.py]: my intention here was to test the method associated with the event, not to test the event trigger machinery.. for me this is sufficient to test how the method behave.

File: src/brasil/gov/vlibrasnews/subscribers.py [hunk 11]
```diff
@@ -45,161 +73,242 @@ def _validate(context, token):
...
-            context.absolute_url(), e.message))
-        return False
+        raise VLibrasNewsError(DELETE_ERROR, e.message)
+
     if response.status_code != 200:
-        logger.error(u'DELETE - {0}: {1} - {2}'.format(
-            context.absolute_url(), response.status_code, response.reason))
-        return False
+        log = u'Status code {0} ({1})'.format(response.status_code, response.reason)
+        raise VLibrasNewsError(DELETE_ERROR, log)
```
Comment (rodfersou) [inline src/brasil/gov/vlibrasnews/subscribers.py:311]: During the video creation the API return status **102**.

**Pull request merged**